stablehlo/dialect/TypeInference.cpp

```diff
@@ -3366,6 +3368,80 @@ LogicalResult verifyCollectivePermuteOp(
   return success();
 }
 
+LogicalResult verifyCompositeOp(std::optional<Location> loc, Operation* op,
+                                StringRef name, StringRef decomposition) {
+  // composite_c1
+  SmallVector<StringRef> parts;
+  name.split(parts, '.');
+  if (
+      // Name must have at least 2 non-empty parts separated by periods.
+      parts.size() < 2 ||
+      llvm::any_of(parts, [](auto part) { return part.empty(); }) ||
+      // First part's first character must be alpha, and subsequent characters
+      // must be alpha, digit or undescore.
+      !isalpha(parts[0][0]) ||
+      !llvm::all_of(
+          llvm::drop_begin(parts[0]),
+          [](char c) { return isalpha(c) || isdigit(c) || c == '_'; }) ||
+      // Subsequent parts must be all alpha, digit, underscore or dollar sign.
+      !llvm::all_of(llvm::drop_begin(parts), [](StringRef part) {
+        return llvm::all_of(part, [](char c) {
+          return isalpha(c) || isdigit(c) || c == '_' || c == '$';
+        });
+      })) {
+    return emitOptionalError(loc,
+                             "name must be a valid namespaced op name, i.e. it",
+                             "must match the following regular expression: ",
+                             "[a-zA-Z][a-zA-Z0-9_]*([.][a-zA-Z0-9_$]+)+, e.g. "
+                             "\"my_namespace.my_op\"");
+  }
+
+  // composite_c2
+  auto module = op->getParentOfType<mlir::ModuleOp>();
+  auto decomp = module.lookupSymbol<mlir::func::FuncOp>(decomposition);
```

Comment:
Let's match the CallOp impl:

```
  FuncOp fn = symbolTable.lookupNearestSymbolFrom<FuncOp>(*this, fnAttr);
  if (!fn)
    return emitOpError() <<  ...
```

Reply:
Unfortunately, this requires implementing `SymbolUserOpInterface`, which I'm not sure we should do. Let's merge the discussion into that thread: https://github.com/openxla/stablehlo/pull/2024#discussion_r1513563283